_posts/2021-03-21-clean-unit-tests-with-mockito.md

```diff
@@ -0,0 +1,440 @@
+    // Test Case Two
+  }
+```
+
+### Write Test Cases Independent
+
+Don't expected test cases to be always executed in the same order! Write everything that belongs to our test case into
+the test method, so that a test case can be executed alone in our IDE or all together within our CI!
+This mistake often comes together with recycled mocks. At least now our alarm bell should ring, and it's time to
+reconsider our test case! Especially, if we need to call `reset()` in a test procedure, which is considered as code
+smell by Mockito.
+
+### Don't Mock Collections nor Values
+
+Mockito is a framework to mock objects with behaviour, that can be declared at the beginning of our test.
+It is common to have *Data Transfer Objects* so called DTOs. The intent of such an DTO is, as its name says, to 
+transport data from a source to a destination. In order to retrieve this data from the object, we could declare
+the behaviour of each getter. Albeit this is possible, we should better use real values and set them to the DTO.
+The same rule applies for collections too, since they are container for values as well.
+
+### Testing Error Handling with Mockito
+
+```java
+Mockito.when(cityRepository.find(expected.getId())).thenThrow(RuntimeException.class);
+```
+
+Mockito comes with a built-in mechanism to test our error handling. Instead of declaring a return value, advise Mockito
+to throw the expected exception. In case we throw checked exceptions, the compiler doesn't let we to throw checked
+exceptions, that are not declared on the method.
+
+### Mocking `void` Methods
+
+```java
+// Causes a compiler error
+Mockito.when(cityRepository.delete()).thenThrow(RuntimeException.class);
+```
+
+In case we want to declare a special behaviour for `void` methods, we must change our approach. The compiler doesn't
+like void methods in brackets, since they have no arguments. Change our approach to following:
+
+```java
+Mockito.doThrow(RuntimeException.class).when(cityRepository).delete(expected);
+```
+
+### Verify Method Calls
+
+```java
+Mockito.verify(cityRepository,Mockito.times(1)).delete(expected);
+```
+
+We can verify how many times a mock was called by simply use the built-in `verify` method. If the condition is not met,
+our test case will fail. This is extremely handy for algorithms or similar processes. There are other predefined
+verification modes such as `atLeastOnce` or `never` already present and ready to use!
+
+## Mockito FAQ
+
+In this section we want to point out important things which are nice to know.
+
+* *What types can I mock?* Mockito allows us to mock not only interfaces but also concrete classes.
+* *What is returned, if I don't declare a mock's behaviour?* Mockito returns `null` for reference objects, and the
+  default values for primitive data types (for example `0` for `int` and `false` for `boolean`)
+* *How many times Mockito returns a previously declared value?* Mockito returns always the same value, regardless of how
```

Comment:
That's not quite true. You can call `when()` multiple times and then the mock returns the values in the order of the `when()` calls.

Reply:
I disagree with that point. I created a test to prove that this is correct. Maybe we misunderstood?